src/storage/src/source/mysql.rs

```diff
@@ -306,57 +308,111 @@ pub(crate) fn pack_mysql_row(
-            },
+            Ok(datum) => datum,
         };
         packer.push(datum);
     }
 
     Ok(Ok(row_container.clone()))
 }
 
+fn mysql_val_to_datum<'a>(
+    value: Value,
+    col_desc: &MySqlColumnDesc,
+    temp_strs: &'a mut Vec<String>,
+    temp_bytes: &'a mut Vec<Vec<u8>>,
+) -> Result<Datum<'a>, anyhow::Error> {
+    Ok(match value {
+        Value::NULL => {
+            if col_desc.column_type.nullable {
+                Datum::Null
+            } else {
+                Err(anyhow::anyhow!(
+                    "received a null value in a non-null column".to_string(),
+                ))?
+            }
+        }
+        value => match &col_desc.column_type.scalar_type {
+            ScalarType::Bool => Datum::from(from_value_opt::<bool>(value)?),
+            ScalarType::UInt16 => Datum::from(from_value_opt::<u16>(value)?),
+            ScalarType::Int16 => Datum::from(from_value_opt::<i16>(value)?),
+            ScalarType::UInt32 => Datum::from(from_value_opt::<u32>(value)?),
+            ScalarType::Int32 => Datum::from(from_value_opt::<i32>(value)?),
+            ScalarType::UInt64 => Datum::from(from_value_opt::<u64>(value)?),
+            ScalarType::Int64 => Datum::from(from_value_opt::<i64>(value)?),
+            ScalarType::Float32 => Datum::from(from_value_opt::<f32>(value)?),
+            ScalarType::Float64 => Datum::from(from_value_opt::<f64>(value)?),
+            ScalarType::Char { length: _ } => {
+                temp_strs.push(from_value_opt::<String>(value)?);
+                Datum::from(temp_strs.last().unwrap().as_str())
+            }
+            ScalarType::VarChar { max_length: _ } => {
+                temp_strs.push(from_value_opt::<String>(value)?);
+                Datum::from(temp_strs.last().unwrap().as_str())
+            }
+            ScalarType::String => {
+                temp_strs.push(from_value_opt::<String>(value)?);
+                Datum::from(temp_strs.last().unwrap().as_str())
+            }
+            ScalarType::Bytes => {
+                let data = from_value_opt::<Vec<u8>>(value)?;
+                temp_bytes.push(data);
+                Datum::from(temp_bytes.last().unwrap().as_slice())
+            }
+            ScalarType::Date => {
+                let date = Date::try_from(from_value_opt::<chrono::NaiveDate>(value)?)?;
+                Datum::from(date)
+            }
+            ScalarType::Timestamp { precision: _ } => {
+                // Timestamps are encoded as different mysql_common::Value types depending on
+                // whether they are from a binlog event or a query, and depending on which
+                // mysql timestamp version is used. We handle those cases here
+                // https://github.com/blackbeam/rust_mysql_common/blob/48ed958051d9a3df1ed1bec3ca0f893b2a18dcc4/src/binlog/value.rs#L87-L155
+                // https://github.com/blackbeam/rust_mysql_common/blob/48ed958051d9a3df1ed1bec3ca0f893b2a18dcc4/src/value/mod.rs#L332
```

Comment:
Maybe link to the version tag we have in our Cargo.lock instead of the commit?
```suggestion
                // https://github.com/blackbeam/rust_mysql_common/blob/v0.31.0/src/binlog/value.rs#L87-L155
                // https://github.com/blackbeam/rust_mysql_common/blob/v0.31.0/src/value/mod.rs#L332
```